src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs

```diff
@@ -5807,17 +5851,26 @@ private void ClearAnalysisCaches()
         /// </remarks>
         private static BinaryAnalysisResult GetCmdletsFromBinaryModuleImplementation(string path, ManifestProcessingFlags manifestProcessingFlags, out Version assemblyVersion)
         {
-            Tuple<BinaryAnalysisResult, Version> tuple;
+            BinaryModuleCacheEntry cacheEntry;
 
-            lock (s_lockObject)
+            s_binaryAnalysisCache.TryGetValue(path, out cacheEntry);
+
+            // Check if the DLL file has changed since last caching
+            DateTime lastFileWriteTime = DateTime.MinValue;
+            try
             {
-                s_binaryAnalysisCache.TryGetValue(path, out tuple);
+                lastFileWriteTime = new FileInfo(path).LastWriteTimeUtc;
```

Comment:
I based the `try`/`catch` thing here on [this](https://github.com/PowerShell/PowerShell/blob/5a326135ec2598cb430d4a1c8cf1533631161361/src/System.Management.Automation/engine/Modules/AnalysisCache.cs#L567) example in `AnalysisCache.cs`.

I don't know if we can trust the path and don't want to present an opportunity to crash PowerShell.

I didn't really want to use exceptions for what is essentially control flow logic, but it seemed more efficient than testing the path and then getting the writetime.

Also, you might note that I've used `LastWriteTimeUtc` unlike in `AnalysisCache` -- which is because I don't want the cache to break because the user changes their clock. (But if my assumption about how that works is wrong, can easily change it too).